fix[ux]: remove side effects in `compare_type` for bytestrings

tserg opened the pull request:
### What I did

Removed side effects in `_Bytestring.compare_type()`.

### How I did it

The implementation of `compare_type` for bytestrings are now simplified to match the top-level definition of `compare_type`: `compare_type should have the meaning: "an expr of type <other> can be assigned to an expr of type <self>."`.

For interface functions that return bytestring types and are imported via JSON, the bytestring types are initialized to zero. To make these functions identifiable, a `returns_abi_bytestring` attribute is added to `FunctionDef` types. Now, when these JSON-imported functions are called and the return value has a concrete bytestring type with a defined length, we create a new copy of the function type and overwrite the return type of the function type during annotation in `visit_Call`. This lets different calls of the same interface function have different lengths for the same bytestring type.

### How to verify it

### Commit message

```
fix: remove side effects in compare_type for bytestrings
```

### Description for the changelog

Remove side effects in compare_type for bytestrings

### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://render.fineartamerica.com/images/rendered/default/greeting-card/images/artworkimages/medium/2/3-sea-otter-and-sleeping-pup-suzi-eszterhas.jpg?&targetx=-25&targety=0&imagewidth=750&imageheight=500&modelwidth=700&modelheight=500&backgroundcolor=2C2A24&orientation=0)

tserg committed:
try what happens

tserg committed:
minor fix

tserg committed:
try fix

tserg committed:
fix lint

tserg commented:
Last failing test may be resolved by https://github.com/vyperlang/vyper/pull/3375

tserg committed:
update comment

tserg committed:
handle nested bytestrings

tserg committed:
fix lint

tserg committed:
try aggressive overriding of contract function return type

tserg committed:
undo changes

tserg committed:
undo changes

charles-cooper commented:
i'm not sure it's such an easy fix. the original intent of this implementation (as far as i can tell) is to deal with progressive widening of abi types. for instance,

```
import jsonabi as SomeInterface  # say one function t() which returns `bytes`

...

@external
def foo():
  x: Bytes[9] = SomeInterface.t()  # t.return_type.min_length is 9
  y: Bytes[10] = SomeInterface.t()  # t.return_type.min_length is now 10
```

tserg committed:
add widening test

tserg committed:
Merge branch 'master' of https://github.com/vyperlang/vyper into fix/bytestring_compare_type

tserg committed:
check for propagated type

tserg committed:
widen bytestring by deriving larger type

tserg commented:
> i'm not sure it's such an easy fix. the original intent of this implementation (as far as i can tell) is to deal with progressive widening of abi types. for instance,
> 
> ```vyper
> import jsonabi as SomeInterface  # say one function t() which returns `bytes`
> 
> ...
> 
> @external
> def foo():
>   x: Bytes[9] = SomeInterface.t()  # t.return_type.min_length is 9
>   y: Bytes[10] = SomeInterface.t()  # t.return_type.min_length is now 10
> ```

I have added this to the test cases. 

From what I understand, the changes made to `_BytestringT.compare_type` will allow the proper type to be derived from the type derivation helpers during annotation, rather than setting directly during typechecking before annotation.

The only other corner case is with JSON ABI imports, and other than the need to override the initial 0-length `BytesT` or `StringT` when the interface function is called, I think the compiler already handles progressive widening as is.

charles-cooper commented:
i see. in that case we may be able to simplify the whole min_length thing in the type.

tserg committed:
remove _min_length; add _is_literal

tserg committed:
update comment

tserg committed:
fix formatting

tserg committed:
improve comment

tserg committed:
use length instead of _length

tserg commented:
I think `slice` can be refactored since `length_literal` cannot be `None` if it must be a constant.
https://github.com/vyperlang/vyper/blob/9cc56b618a54e6144829d6c056feb43e00c9c5fe/vyper/builtins/functions.py#L311-L329

Should I do it here or in a follow-up PR?

tserg committed:
add comment

tserg committed:
fix lint

charles-cooper commented:
for some reason i thought this was merged already. @tserg can you fix the merge conflicts and we can review this again?

tserg committed:
Merge branch 'master' of https://github.com/vyperlang/vyper into fix/bytestring_compare_type

tserg committed:
fix literal cmp

tserg committed:
uncomment

tserg committed:
fix wip

tserg committed:
Merge branch 'master' of https://github.com/vyperlang/vyper into fix/bytestring_compare_type

tserg committed:
fix test

tserg commented on vyper/semantics/analysis/utils.py:
```diff
@@ -416,7 +416,7 @@ def _is_empty_list(node):
 def _is_type_in_list(obj, types_list):
     # check if a type object is in a list of types
-    return any(i.compare_type(obj) for i in types_list)
+    return any(obj.compare_type(i) for i in types_list)
```
Haven't figured out why the current form would throw for this valid contract, but compile once it is flipped:
```
@external
def foo() -> String[33]:
    for x in ["hello", ",", "world"]:
        if x == ",":
            return x
    return ""
```

tserg committed:
fix any syntax

tserg committed:
fix comment

tserg committed:
rename helper; change semantics

tserg committed:
fix lint

tserg committed:
add is_from_abi attr

tserg committed:
remove set_length fn

tserg committed:
fix slice

tserg committed:
fix lint

tserg committed:
remove is_literal

tserg committed:
fix compare_type

tserg committed:
revert overwriting of ContractFunctionT return typ

tserg committed:
add comment

tserg committed:
fix lint

charles-cooper committed:
wip - review

charles-cooper committed:
fix lint

charles-cooper committed:
fix some return types

tserg committed:
fix return type for call stmts

charles-cooper committed:
Merge branch 'master' into fix/bytestring_compare_type

charles-cooper committed:
rename to _any_compare_type

charles-cooper committed:
fix lint

charles-cooper committed:
rename fetch_call_return to get_return_type

remove InvalidType exception from validate_expected_type
rewrite _Bytestring.compare_type for clarity

charles-cooper committed:
fix some signatures

tserg committed:
Merge branch 'master' of https://github.com/vyperlang/vyper into fix/bytestring_compare_type

tserg committed:
Merge branch 'fix/bytestring_compare_type' of https://github.com/tserg/vyper into fix/bytestring_compare_type

tserg committed:
fix lint

tserg committed:
fix more lint

tserg committed:
use Optional

tserg committed:
revert return analysis

tserg committed:
fix some tests

DanielSchiavini commented:
Is this a duplicate of https://github.com/charles-cooper/vyper/pull/17?

tserg committed:
fix more tests

tserg committed:
revert circular call in infer_arg_types

tserg committed:
fix lint

tserg commented:
> Is this a duplicate of https://github.com/charles-cooper/vyper/pull/17 and/or https://github.com/charles-cooper/vyper/pull/18?

Yes, it does look like they are addressing the same issue.

tserg committed:
fix more tests

tserg committed:
fix lint

tserg committed:
fix import

tserg committed:
Merge branch 'master' of https://github.com/vyperlang/vyper into fix/bytestring_compare_type

tserg committed:
fix lint

tserg committed:
fix convert

tserg committed:
fix lint

tserg committed:
fix typo

tserg committed:
fix another typo

tserg committed:
fix minmax

charles-cooper committed:
Merge branch 'master' into fix/bytestring_compare_type

charles-cooper committed:
fix a test

tserg committed:
Merge branch 'fix/bytestring_compare_type' of https://github.com/tserg/vyper into fix/bytestring_compare_type

tserg committed:
Merge branch 'master' of https://github.com/vyperlang/vyper into fix/bytestring_compare_type

tserg committed:
fix derivation of return type for external call

tserg committed:
fix bytes assign check

Sporarum commented:
Very cool work, I ended up doing something similar in my ongoing implementation of `Bytes[INF]` which turns out to be the "correct" type for JSON abi `bytes`. [VIP](https://github.com/vyperlang/vyper/issues/4856) [snapshot of implementation](https://github.com/Sporarum/vyper/pull/1)
Once that's finished, I'll look back at this PR, and see if all its tests pass

Sporarum commented:
In #4967 I ended up implementing the same tests (and fixing the issues) so this is no longer needed

Thank you for the contribution however !
